Handle semver-style version tags with prerelease and buildmetadata
information as used for the development releases, and gracefully deal
with repositories in a "detached HEAD" state.  The repository will
need a TVIPS-TOOLS-VERSION-FILE file in the tvips directory, otherwise
the patched tree will be flagged as dirty.  The version file cannot be
named VERSION, because that will clash with a system-provided header
on case-insensitive macOS filesystems.
--- tvips/CMakeLists.txt.orig
+++ tvips/CMakeLists.txt
@@ -116,14 +116,14 @@ CHECK_FUNCTION_EXISTS(getprogname HAVE_GETPROGNAME)
 CHECK_FUNCTION_EXISTS(setprogname HAVE_SETPROGNAME)
 
 
-# Add a rule to generate the VERSION file in the source directory,
-# which is required by the dist target.
+# Add a rule to generate the TVIPS-TOOLS-VERSION-FILE file in the
+# source directory, which is required by the dist target.
 add_custom_command(
   COMMAND ${CMAKE_COMMAND}
-    "-DPATH=${PROJECT_SOURCE_DIR}/VERSION"
+    "-DPATH=${PROJECT_SOURCE_DIR}/TVIPS-TOOLS-VERSION-FILE"
     "-DWRITE:BOOL=true"
     -P "${PROJECT_SOURCE_DIR}/version.cmake"
-  OUTPUT "${PROJECT_SOURCE_DIR}/VERSION"
+  OUTPUT "${PROJECT_SOURCE_DIR}/TVIPS-TOOLS-VERSION-FILE"
   VERBATIM)
 
 
@@ -286,9 +286,9 @@ install(
 # the source directory.  In order to include the pre-built
 # documentation as well as the version information file, the dist
 # target must depend on the documentation files as well as
-# ${PROJECT_SOURCE_DIR}/VERSION.  The documentation files are copied
-# to ${PROJECT_SOURCE_DIR} and removed once the package_source target
-# has completed.
+# ${PROJECT_SOURCE_DIR}/TVIPS-TOOLS-VERSION-FILE.  The documentation
+# files are copied to ${PROJECT_SOURCE_DIR} and removed once the
+# package_source target has completed.
 #
 # Determine the version, patch level, and the source status as in
 # config.cmake.  ${version_branch}, which is the last component of the
@@ -310,17 +310,17 @@ add_custom_target(dist
     "${PROJECT_SOURCE_DIR}/tiff2smv.1"
     "${PROJECT_SOURCE_DIR}/tiff2smv.pdf"
     "${PROJECT_SOURCE_DIR}/tvips2smv.1"
-    "${PROJECT_SOURCE_DIR}/VERSION"
+    "${PROJECT_SOURCE_DIR}/TVIPS-TOOLS-VERSION-FILE"
   DEPENDS "${PROJECT_BINARY_DIR}/tiff2smv.1"
   DEPENDS "${PROJECT_BINARY_DIR}/tiff2smv.pdf"
   DEPENDS "${PROJECT_BINARY_DIR}/tvips2smv.1"
-  DEPENDS "${PROJECT_SOURCE_DIR}/VERSION"
+  DEPENDS "${PROJECT_SOURCE_DIR}/TVIPS-TOOLS-VERSION-FILE"
   VERBATIM)
 
 execute_process(
   COMMAND ${CMAKE_COMMAND}
     "-DOUTPUT=major"
-    "-DPATH=${PROJECT_SOURCE_DIR}/VERSION"
+    "-DPATH=${PROJECT_SOURCE_DIR}/TVIPS-TOOLS-VERSION-FILE"
     -P "${PROJECT_SOURCE_DIR}/version.cmake"
   OUTPUT_VARIABLE CPACK_PACKAGE_VERSION_MAJOR
   OUTPUT_STRIP_TRAILING_WHITESPACE)
@@ -328,7 +328,7 @@ execute_process(
 execute_process(
   COMMAND ${CMAKE_COMMAND}
     "-DOUTPUT=minor"
-    "-DPATH=${PROJECT_SOURCE_DIR}/VERSION"
+    "-DPATH=${PROJECT_SOURCE_DIR}/TVIPS-TOOLS-VERSION-FILE"
     -P "${PROJECT_SOURCE_DIR}/version.cmake"
   OUTPUT_VARIABLE CPACK_PACKAGE_VERSION_MINOR
   OUTPUT_STRIP_TRAILING_WHITESPACE)
@@ -336,7 +336,7 @@ execute_process(
 execute_process(
   COMMAND ${CMAKE_COMMAND}
     "-DOUTPUT=patch"
-    "-DPATH=${PROJECT_SOURCE_DIR}/VERSION"
+    "-DPATH=${PROJECT_SOURCE_DIR}/TVIPS-TOOLS-VERSION-FILE"
     -P "${PROJECT_SOURCE_DIR}/version.cmake"
   OUTPUT_VARIABLE CPACK_PACKAGE_VERSION_PATCH
   OUTPUT_STRIP_TRAILING_WHITESPACE)
@@ -344,7 +344,7 @@ execute_process(
 execute_process(
   COMMAND ${CMAKE_COMMAND}
     "-DOUTPUT=tweak"
-    "-DPATH=${PROJECT_SOURCE_DIR}/VERSION"
+    "-DPATH=${PROJECT_SOURCE_DIR}/TVIPS-TOOLS-VERSION-FILE"
     -P "${PROJECT_SOURCE_DIR}/version.cmake"
   OUTPUT_VARIABLE version_tweak
   OUTPUT_STRIP_TRAILING_WHITESPACE)
@@ -352,7 +352,7 @@ execute_process(
 execute_process(
   COMMAND ${CMAKE_COMMAND}
     "-DOUTPUT=branch"
-    "-DPATH=${PROJECT_SOURCE_DIR}/VERSION"
+    "-DPATH=${PROJECT_SOURCE_DIR}/TVIPS-TOOLS-VERSION-FILE"
     -P "${PROJECT_SOURCE_DIR}/version.cmake"
   OUTPUT_VARIABLE version_branch
   OUTPUT_STRIP_TRAILING_WHITESPACE)
@@ -363,7 +363,7 @@ string(REGEX REPLACE "[^-@0-9a-zA-Z]" "_" version_branch "${version_branch}")
 execute_process(
   COMMAND ${CMAKE_COMMAND}
     "-DOUTPUT=status"
-    "-DPATH=${PROJECT_SOURCE_DIR}/VERSION"
+    "-DPATH=${PROJECT_SOURCE_DIR}/TVIPS-TOOLS-VERSION-FILE"
     -P "${PROJECT_SOURCE_DIR}/version.cmake"
   OUTPUT_VARIABLE version_status
   OUTPUT_STRIP_TRAILING_WHITESPACE)
--- tvips/config.cmake.orig
+++ tvips/config.cmake
@@ -13,7 +13,7 @@
 execute_process(
   COMMAND ${CMAKE_COMMAND}
     "-DOUTPUT=major"
-    "-DPATH=${PROJECT_SOURCE_DIR}/VERSION"
+    "-DPATH=${PROJECT_SOURCE_DIR}/TVIPS-TOOLS-VERSION-FILE"
     -P "${PROJECT_SOURCE_DIR}/version.cmake"
   OUTPUT_VARIABLE PACKAGE_VERSION_MAJOR
   OUTPUT_STRIP_TRAILING_WHITESPACE)
@@ -21,7 +21,7 @@ execute_process(
 execute_process(
   COMMAND ${CMAKE_COMMAND}
     "-DOUTPUT=minor"
-    "-DPATH=${PROJECT_SOURCE_DIR}/VERSION"
+    "-DPATH=${PROJECT_SOURCE_DIR}/TVIPS-TOOLS-VERSION-FILE"
     -P "${PROJECT_SOURCE_DIR}/version.cmake"
   OUTPUT_VARIABLE PACKAGE_VERSION_MINOR
   OUTPUT_STRIP_TRAILING_WHITESPACE)
@@ -29,7 +29,7 @@ execute_process(
 execute_process(
   COMMAND ${CMAKE_COMMAND}
     "-DOUTPUT=patch"
-    "-DPATH=${PROJECT_SOURCE_DIR}/VERSION"
+    "-DPATH=${PROJECT_SOURCE_DIR}/TVIPS-TOOLS-VERSION-FILE"
     -P "${PROJECT_SOURCE_DIR}/version.cmake"
   OUTPUT_VARIABLE PACKAGE_VERSION_PATCH
   OUTPUT_STRIP_TRAILING_WHITESPACE)
@@ -37,15 +37,31 @@ execute_process(
 execute_process(
   COMMAND ${CMAKE_COMMAND}
     "-DOUTPUT=tweak"
-    "-DPATH=${PROJECT_SOURCE_DIR}/VERSION"
+    "-DPATH=${PROJECT_SOURCE_DIR}/TVIPS-TOOLS-VERSION-FILE"
     -P "${PROJECT_SOURCE_DIR}/version.cmake"
   OUTPUT_VARIABLE PACKAGE_VERSION_TWEAK
   OUTPUT_STRIP_TRAILING_WHITESPACE)
 
+execute_process(
+  COMMAND ${CMAKE_COMMAND}
+    "-DOUTPUT=prerelease"
+    "-DPATH=${PROJECT_SOURCE_DIR}/TVIPS-TOOLS-VERSION-FILE"
+    -P "${PROJECT_SOURCE_DIR}/version.cmake"
+  OUTPUT_VARIABLE PACKAGE_VERSION_PRERELEASE
+  OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+execute_process(
+  COMMAND ${CMAKE_COMMAND}
+    "-DOUTPUT=buildmetadata"
+    "-DPATH=${PROJECT_SOURCE_DIR}/TVIPS-TOOLS-VERSION-FILE"
+    -P "${PROJECT_SOURCE_DIR}/version.cmake"
+  OUTPUT_VARIABLE PACKAGE_VERSION_BUILDMETADATA
+  OUTPUT_STRIP_TRAILING_WHITESPACE)
+
 execute_process(
   COMMAND ${CMAKE_COMMAND}
     "-DOUTPUT=branch"
-    "-DPATH=${PROJECT_SOURCE_DIR}/VERSION"
+    "-DPATH=${PROJECT_SOURCE_DIR}/TVIPS-TOOLS-VERSION-FILE"
     -P "${PROJECT_SOURCE_DIR}/version.cmake"
   OUTPUT_VARIABLE PACKAGE_VERSION_BRANCH
   OUTPUT_STRIP_TRAILING_WHITESPACE)
@@ -53,7 +69,7 @@ execute_process(
 execute_process(
   COMMAND ${CMAKE_COMMAND}
     "-DOUTPUT=commit"
-    "-DPATH=${PROJECT_SOURCE_DIR}/VERSION"
+    "-DPATH=${PROJECT_SOURCE_DIR}/TVIPS-TOOLS-VERSION-FILE"
     -P "${PROJECT_SOURCE_DIR}/version.cmake"
   OUTPUT_VARIABLE PACKAGE_VERSION_COMMIT
   OUTPUT_STRIP_TRAILING_WHITESPACE)
@@ -61,7 +77,7 @@ execute_process(
 execute_process(
   COMMAND ${CMAKE_COMMAND}
     "-DOUTPUT=status"
-    "-DPATH=${PROJECT_SOURCE_DIR}/VERSION"
+    "-DPATH=${PROJECT_SOURCE_DIR}/TVIPS-TOOLS-VERSION-FILE"
     -P "${PROJECT_SOURCE_DIR}/version.cmake"
   OUTPUT_VARIABLE PACKAGE_VERSION_STATUS
   OUTPUT_STRIP_TRAILING_WHITESPACE)
--- tvips/config.h.in.orig
+++ tvips/config.h.in
@@ -37,6 +37,10 @@
  */
 #define PACKAGE_VERSION_BRANCH "${PACKAGE_VERSION_BRANCH}"
 
+/* Build metadata identifier, if any
+ */
+#define PACKAGE_VERSION_BUILDMETADATA "${PACKAGE_VERSION_BUILDMETADATA}"
+
 /* The integer major version of the package
  */
 #define PACKAGE_VERSION_MAJOR ${PACKAGE_VERSION_MAJOR}
@@ -49,6 +53,10 @@
  */
 #define PACKAGE_VERSION_PATCH ${PACKAGE_VERSION_PATCH}
 
+/* Pre-release identifier, if any
+ */
+#define PACKAGE_VERSION_PRERELEASE "${PACKAGE_VERSION_PRERELEASE}"
+
 /* The number of additional commits on top of the tagged object
  */
 #define PACKAGE_VERSION_TWEAK ${PACKAGE_VERSION_TWEAK}
--- tvips/util.c.orig
+++ tvips/util.c
@@ -151,20 +151,26 @@ version()
                 PACKAGE_NAME);
 
     } else {
-        fprintf(stdout, "%s (%s) %d.%d.%d\n",
+        fprintf(stdout, "%s (%s) %d.%d.%d%s%s\n",
                 getprogname(),
                 PACKAGE_NAME,
                 PACKAGE_VERSION_MAJOR,
                 PACKAGE_VERSION_MINOR,
-                PACKAGE_VERSION_PATCH);
+                PACKAGE_VERSION_PATCH,
+                strlen(PACKAGE_VERSION_PRERELEASE) > 0
+                ? "-"  PACKAGE_VERSION_PRERELEASE
+                : "",
+                strlen(PACKAGE_VERSION_BUILDMETADATA) > 0
+                ? "+"  PACKAGE_VERSION_BUILDMETADATA
+                : "");
 
         if (PACKAGE_VERSION_TWEAK > 0 ||
-            strcasecmp(PACKAGE_VERSION_BRANCH, "master") != 0 ||
             strcasecmp(PACKAGE_VERSION_STATUS, "dirty") == 0) {
 
             fprintf(stdout,  "    %s%s",
                     PACKAGE_VERSION_BRANCH,
-                    strlen(PACKAGE_VERSION_COMMIT) > 0
+                    strpbrk(PACKAGE_VERSION_BRANCH, " \t") == NULL
+                    && strlen(PACKAGE_VERSION_COMMIT) > 0
                     ? "." PACKAGE_VERSION_COMMIT
                     : "");
             if (strcasecmp(PACKAGE_VERSION_STATUS, "dirty") == 0) {
--- tvips/version.cmake.orig
+++ tvips/version.cmake
@@ -28,7 +28,7 @@ else()
   find_program(git git)
 
   execute_process(
-    COMMAND "${git}" describe --abbrev=7 --match "v[0-9]*" --dirty=-dirty
+    COMMAND "${git}" describe --dirty --long --match "v[0-9]*" --tags
     ERROR_VARIABLE errors
     OUTPUT_VARIABLE descriptions
     OUTPUT_STRIP_TRAILING_WHITESPACE
@@ -96,13 +96,26 @@ elseif("${OUTPUT}" STREQUAL "patch")
   string(REGEX REPLACE "^v[0-9]+\\.[0-9]+\\." "" patch "${patch}")
   execute_process(COMMAND ${CMAKE_COMMAND} -E echo "${patch}")
 
+elseif("${OUTPUT}" STREQUAL "prerelease")
+  string(REGEX MATCH "^v[0-9]+\\.[0-9]+\\.[0-9]+-[^0-9][^+-]*" prerelease "${vi}")
+  string(REGEX REPLACE "^v[0-9]+\\.[0-9]+\\.[0-9]+-" "" prerelease "${prerelease}")
+  execute_process(COMMAND ${CMAKE_COMMAND} -E echo "${prerelease}")
+
+elseif("${OUTPUT}" STREQUAL "buildmetadata")
+  string(REGEX MATCH "^v[0-9]+\\.[0-9]+\\.[0-9]+-[^0-9][^+-]*\\+[^-]+" buildmetadata "${vi}")
+  string(REGEX REPLACE "^v[0-9]+\\.[0-9]+\\.[0-9]+-[^0-9][^+-]*\\+" "" buildmetadata "${buildmetadata}")
+  execute_process(COMMAND ${CMAKE_COMMAND} -E echo "${buildmetadata}")
+
 elseif("${OUTPUT}" STREQUAL "commit")
   string(REGEX MATCH "^v[\\.0-9-]+-g[0-9a-f]+" commit "${vi}")
   string(REGEX REPLACE "^v[\\.0-9-]+-g" "" commit "${commit}")
   execute_process(COMMAND ${CMAKE_COMMAND} -E echo "${commit}")
 
 elseif("${OUTPUT}" STREQUAL "branch")
-  string(REGEX MATCH "-[^-]+$" branch "${vi}")
+  string(REGEX MATCH "-\\(.+\\)$" branch "${vi}")
+  if("${branch}" STREQUAL "")
+    string(REGEX MATCH "-[^-]+$" branch "${vi}")
+  endif()
   string(REGEX REPLACE "^-" "" branch "${branch}")
   execute_process(COMMAND ${CMAKE_COMMAND} -E echo "${branch}")
 
